docs/rfcs/0003-ecommerce-models.md

```diff
@@ -0,0 +1,136 @@
+## Title for RFC
+
+ecommerce workflow
+
+#### Abstract
+
+This describes the various states of ecommerce models and where and when we can and cannot
+modify data.
+
+#### Product
+
+A `Product` has a `price`, `description`, and a `GenericForeignKey` `content_object` which can
+refer to any Django model instance. An admin user should be able to edit
+`description` or `price` without affecting any previous purchase, so we need to also store
+this separately when the purchase is made.
+
+#### Basket
+
+A `Basket` has foreign keys to an `Order` and a `User`. The `User` foreign key must always exist.
+
+A `Basket` may be editable or not editable, depending on the `editable` boolean field. In general
+a `Basket` starts off editable so that the user can add products or a coupon. Then it becomes not
+editable once it is passed to CyberSource for payment.
+
+When a `Basket` is editable it should not have a link to an `Order`, and when it is not editable
+it should have a link to an `Order`. This makes the `editable` flag a little redundant but this
+makes the intent more explicit, and it provides flexibility in the future if we want to change
+the logic around when `Order` is created and linked to a `Basket`.
+
+There should be at most one `Basket` for a user which is editable.
+This is the user's active basket. If there is no editable `Basket` for the user,
+it should be created once the user adds a product or a coupon.
+
+A basket may have at most one coupon. See `CouponBasket` below.
+
+##### Edge cases
+
+Once a user checks out and is directed to CyberSource, the basket stops being editable even if
+they haven't actually paid yet. If the user tries to go back they will find their basket to be empty
+and they will have to add their items again.
+
+Fixing this complicates the logic significantly, and the UI does not really have a basket, so I'm
+ignoring this edge case for now.
+
+#### Line
+
+A `Line` is an item in a `Basket`. It has a foreign key to the `Basket` and to `Product`. There
+are also fields for `price_per_unit`, `total_price`, `quantity` and `description`.
+
+While the basket is editable the price fields and description field will be null because this
```

Comment:
An alternative approach would be to store the editable `Product` fields on another table (e.g. `ProductVersion`) that is append-only with immutable rows, which whenever the product is edited a new row with the latest values is written. `Line` could then reference both `Product` and then either a nullable `ForeignKey` to `ProductVersion` where if `NOT NULL` references the `ProductVersion` at which the `Product` was purchased or if `IS NULL` dynamically uses the latest `ProductVersion`. This would have the benefit of not needing to keep multiple copies of the data on different tables and also avoid the need to keep both data models and sync code up-to-date with each other and instead rely on the data model.

At that point you'd really probably only need quantity since the other fields are computable.

Reply:
I think that's a good idea, since then we'd also have a complete history of pricing for any particular product

Reply:
@rhysyngsun Do you think it makes sense to make `ProductVersion` an audit table of `Product`? The fields would be stored as JSON but maybe that's fine since we can still filter on the data with the JSONField.

Reply:
@noisecapella depends, do we imagine any tooling around this data needing more traditionally structured records? (e.g. how easy is `JSONField` filtering in django-admin? how easy will it be for BI purposes? will it complicate any query writing?)

Reply:
FYI, Max is doing some testing now with MicroMasters around the utility of the audit tables. We have a couple of specific cases in mind, like refunds.

Reply:
I'll implement it as an audit table. I'm pretty sure you can interact with these JSON fields via Django queries or SQL, it just won't be indexed. I don't think indexing matters here since there are going to be very few products compared to coupons, orders and users.

Reply:
Reading over this again this morning, @rhysyngsun's proposal makes the most sense to me. The audit table won't make sense unless there is a table to be audited and it seems like the proposal to have `ProductVersion` be the source of truth makes the most sense here.